lib/eknvfs/ekn-vfs.c

```diff
@@ -0,0 +1,340 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/*
+ * ekn-vfs.c
+ *
+ * Copyright (C) 2016 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Author: Juan Pablo Ugarte <ugarte@endlessm.com>
+ *
+ */
+
+#include "ekn-file.h"
+#include <string.h>
+#include <eos-shard/eos-shard-shard-file.h>
+#include <eos-shard/eos-shard-record.h>
+
+#define EKN_TYPE_VFS (ekn_vfs_get_type ())
+G_DECLARE_FINAL_TYPE (EknVfs, ekn_vfs, EKN, VFS, GVfs)
+
+#define EKN_URI "ekn"
+#define EKN_SCHEME_LEN 6
+
+struct _EknVfs
+{
+  GVfs parent;
+};
+
+typedef struct
+{
+  GSList *shards;         /* EosShardShardFile list */
+
+  GHashTable *extensions; /* (uri, GVfs *) table */
```

Comment:
This hash table is to map existing vfs extension points to the uri they can handle.
In practice there is usually only two, the local vfs and gvfsd, so it might be faster to store them in an array
but I do not think it makes much of a difference